File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/error/PlayerErrorParserImpl.java
package com.github.anrimian.musicplayer.data.controllers.music.error;

import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.google.android.exoplayer2.PlaybackException;
import com.google.android.exoplayer2.source.UnrecognizedInputFormatException;
import com.google.android.exoplayer2.upstream.ContentDataSource;
import com.google.android.exoplayer2.upstream.FileDataSource;

import java.io.FileNotFoundException;

public class PlayerErrorParserImpl implements PlayerErrorParser {

    private final Analytics analytics;

    public PlayerErrorParserImpl(Analytics analytics) {
        this.analytics = analytics;
    }

    @Override
    public ErrorType getErrorType(Throwable throwable) {
        if (throwable instanceof FileNotFoundException) {
            return ErrorType.NOT_FOUND;
        }
        if (throwable instanceof PlaybackException) {
            Throwable cause = throwable.getCause();
            if (cause instanceof ContentDataSource.ContentDataSourceException
                || cause instanceof FileDataSource.FileDataSourceException) {
                Throwable causeOfCause = cause.getCause();
                if (causeOfCause instanceof FileNotFoundException) {
                    return ErrorType.NOT_FOUND;
                }
            }
            if (cause instanceof UnrecognizedInputFormatException) {
                return ErrorType.UNSUPPORTED;
            }
        }
        if (throwable instanceof SecurityException) {
            return ErrorType.IGNORED;
        }
        analytics.processNonFatalError(throwable);
        return ErrorType.UNKNOWN;
    }
}

File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/error/PlayerErrorParserImpl.kt
package com.github.anrimian.musicplayer.data.controllers.music.error

import androidx.annotation.OptIn
import androidx.media3.common.ParserException
import androidx.media3.common.PlaybackException
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.ContentDataSource
import androidx.media3.datasource.FileDataSource
import androidx.media3.exoplayer.source.UnrecognizedInputFormatException
import androidx.media3.exoplayer.upstream.Loader
import com.github.anrimian.musicplayer.data.models.exceptions.CompositionNotFoundException
import com.github.anrimian.musicplayer.data.storage.exceptions.UnavailableMediaStoreException
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics
import com.github.anrimian.musicplayer.domain.interactors.player.PlayerErrorParser
import com.github.anrimian.musicplayer.domain.models.composition.content.AcceptablePlayerException
import com.github.anrimian.musicplayer.domain.models.composition.content.CorruptedMediaFileException
import com.github.anrimian.musicplayer.domain.models.composition.content.LocalSourceNotFoundException
import com.github.anrimian.musicplayer.domain.models.composition.content.NoReadPermissionException
import com.github.anrimian.musicplayer.domain.models.composition.content.RelaunchSourceException
import com.github.anrimian.musicplayer.domain.models.composition.content.TooLargeSourceException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnknownPlayerException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnsupportedSourceException
import com.github.anrimian.musicplayer.domain.models.exceptions.NotAllowedPathException
import java.io.FileNotFoundException

open class PlayerErrorParserImpl(
    private val analytics: Analytics,
): PlayerErrorParser {

    @OptIn(UnstableApi::class)
    override fun parseError(throwable: Throwable): Throwable {
        when (throwable) {
            is RelaunchSourceException -> return throwable
            is FileNotFoundException -> return LocalSourceNotFoundException()
            is CompositionNotFoundException -> return LocalSourceNotFoundException()
            is PlaybackException -> {
                val cause = throwable.cause
                if (cause is Loader.UnexpectedLoaderException) {
                    val relaunchCause = when (val causeOfCause = cause.cause) {
                        is OutOfMemoryError -> TooLargeSourceException(causeOfCause)
                        else -> cause
                    }
                    return RelaunchSourceException(relaunchCause)
                }
                if (cause is ContentDataSource.ContentDataSourceException || cause is FileDataSource.FileDataSourceException) {
                    val causeOfCause = cause.cause
                    if (causeOfCause is FileNotFoundException) {
                        return LocalSourceNotFoundException()
                    }
                }
                if (cause is UnrecognizedInputFormatException) {
                    return UnsupportedSourceException()
                }
                if (cause is ParserException) {
                    return CorruptedMediaFileException()
                }
                if (isTransientCodecError(throwable)) {
                    return RelaunchSourceException(throwable)
                }
            }
            is UnsupportedSourceException -> return throwable
            is NoReadPermissionException -> return throwable
            is UnavailableMediaStoreException -> return AcceptablePlayerException(throwable)
            is NotAllowedPathException -> return throwable
            is UnknownPlayerException,
            is SecurityException,
                -> {
                analytics.processNonFatalError(throwable)
                return RelaunchSourceException(throwable)
            }
        }
        analytics.processNonFatalError(throwable)
        return throwable
    }

    private fun isTransientCodecError(throwable: Throwable): Boolean {
        var cause: Throwable? = throwable
        while (cause != null) {
            if (cause is android.media.MediaCodec.CodecException) {
                return true
            }
            cause = cause.cause
        }
        return false
    }
}